router/batchrouter/asyncdestinationmanager/lytics_bulk_upload/lyticsBulkUpload.go

```diff
@@ -0,0 +1,380 @@
+	// Create a temporary directory using misc.CreateTMPDIR
+	tmpDirPath, err := misc.CreateTMPDIR()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temporary directory: %v", err)
+	}
+
+	// Define a local directory name within the temp directory
+	localTmpDirName := fmt.Sprintf("/%s/", misc.RudderAsyncDestinationLogs)
+
+	// Combine the temporary directory with the local directory name and generate a unique file path
+	path := filepath.Join(tmpDirPath, localTmpDirName, uuid.NewString())
+	csvFilePath := fmt.Sprintf("%v.csv", path)
+
+	// Initialize the CSV writer with the generated file path
+	actionFile, err := createCSVWriter(csvFilePath)
+	if err != nil {
+		return nil, err
+	}
+	defer actionFile.File.Close() // Ensure the file is closed when done
+
+	// Store the CSV file path in the ActionFileInfo struct
+	actionFile.CSVFilePath = csvFilePath
+
+	// Create a scanner to read the existing file line by line
+	existingFile, err := os.Open(existingFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open existing file: %v", err)
+	}
+	defer existingFile.Close()
+
+	scanner := bufio.NewScanner(existingFile)
+	scanner.Buffer(nil, 50000*1024) // Adjust the buffer size if necessary
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		var data Data
+		if err := jsoniter.Unmarshal([]byte(line), &data); err != nil {
+			// Collect the failed job ID
+			actionFile.FailedJobIDs = append(actionFile.FailedJobIDs, data.Metadata.JobID)
+			continue
+		}
+
+		// Calculate the payload size and observe it
+		payloadSizeStat := stats.Default.NewTaggedStat("payload_size", stats.HistogramType,
+			map[string]string{
+				"module":   "batch_router",
+				"destType": b.destName,
+			})
+		payloadSizeStat.Observe(float64(len(data.Message.Properties)))
+
+		// Populate the CSV file and collect success/failure job IDs
+		err := b.PopulateCsvFile(actionFile, streamTraitsMapping, line, data)
+		if err != nil {
+			actionFile.FailedJobIDs = append(actionFile.FailedJobIDs, data.Metadata.JobID)
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error while scanning file: %v", err)
+	}
+
+	// After processing, calculate the final file size
+	fileInfo, err := os.Stat(actionFile.CSVFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve file info: %v", err)
+	}
+	actionFile.FileSize = fileInfo.Size()
+
+	return actionFile, nil
+}
+
+func convertGjsonToStreamTraitMapping(result gjson.Result) []StreamTraitMapping {
+	var mappings []StreamTraitMapping
+
+	// Iterate through the array in the result
+	result.ForEach(func(key, value gjson.Result) bool {
+		mapping := StreamTraitMapping{
+			RudderProperty: value.Get("rudderProperty").String(),
+			LyticsProperty: value.Get("lyticsProperty").String(),
+		}
+		mappings = append(mappings, mapping)
+		return true // Continue iteration
+	})
+
+	return mappings
+}
+
+func (e *LyticsBulkUploader) MakeHTTPRequest(data *HttpRequestData) ([]byte, int, error) {
```

Comment:
https://pkg.go.dev/net/http/httptest use this write tests

Reply:
updated